src/gfx.cpp

```diff
@@ -90,6 +90,155 @@ const char* GfxProc::supportedvideoformats()
     return NULL;
 }
 
+void *GfxProc::threadEntryPoint(void *param)
+{
+    GfxProc* gfxProcessor = (GfxProc*)param;
+    gfxProcessor->loop();
+    return NULL;
+}
+
+void GfxProc::loop()
+{
+    GfxJob *job = NULL;
+    while (!finished)
+    {
+        waiter.init(NEVER);
+        waiter.wait();
+        while (job = requests.pop())
+        {
+            if (finished)
+            {
+                delete job;
+                break;
+            }
+
+            mutex.lock();
+            LOG_debug << "Processing media file: " << job->h;
+
+            // (this assumes that the width of the largest dimension is max)
+            if (readbitmap(NULL, &job->localfilename, dimensions[sizeof dimensions/sizeof dimensions[0]-1][0]))
+            {
+                for (int i = 0; i < job->imagetypes.size(); i++)
+                {
+                    // successively downscale the original image
+                    string* jpeg = new string();
+                    int w = dimensions[job->imagetypes[i]][0];
+                    int h = dimensions[job->imagetypes[i]][1];
+
+                    if (job->imagetypes[i] == PREVIEW && this->w < w && this->h < h )
+                    {
+                        LOG_debug << "Skipping upsizing of preview";
+                        w = this->w;
+                        h = this->h;
+                    }
+
+                    if (!resizebitmap(w, h, jpeg))
+                    {
+                        delete jpeg;
+                        jpeg = NULL;
+                    }
+                    job->images.push_back(jpeg);
+                }
+                freebitmap();
+            }
+            else
+            {
+                for (int i = 0; i < job->imagetypes.size(); i++)
+                {
+                    job->images.push_back(NULL);
+                }
+            }
+
+            mutex.unlock();
+            responses.push(job);
+            client->waiter->notify();
+        }
+    }
+
+    while (job = requests.pop())
+    {
+        delete job;
+    }
+
+    while (job = responses.pop())
+    {
+        for (int i = 0; i < job->imagetypes.size(); i++)
+        {
+            delete job->images[i];
+        }
+        delete job;
```

Comment:
couldn't be this deletion be called before the response is processed in GfxProc::checkevents?